python/lsst/daf/butler/datastores/fileDatastore.py

```diff
@@ -1122,7 +1127,24 @@ def _extractIngestInfo(
     def _prepIngest(self, *datasets: FileDataset, transfer: str | None = None) -> _IngestPrepData:
         # Docstring inherited from Datastore._prepIngest.
         filtered = []
-        for dataset in datasets:
+
+        # Ingest could be given tens of thousands of files. It is not efficient
+        # to check for the existence of every single file (especially if they
+        # are remote URIs) but in some transfer modes the files will be checked
+        # anyhow when they are relocated. For direct or None transfer modes
+        # it is possible to not know if the file is accessible at all.
+        # Therefore limit number of files that will be checked (but always
+        # include the first one).
+        max_checks = 200
+        n_datasets = len(datasets)
+        if n_datasets <= max_checks:
+            check_every_n = 1
+        elif transfer in ("direct", None):
+            check_every_n = int(n_datasets / max_checks + 1)  # +1 so that if n < 100 the answer is 1.
```

Comment:
@ktlim I've asked for a re-review since this change is quite significant. The code coverage doesn't cover the important bit because none of the tests try to ingest large numbers of files. I have verified that @gpdf 's example does only check 200 files out of 155k and that doesn't notable affect the import time.